rfc/RFC_for_CI_roles_extraction.md

```diff
@@ -0,0 +1,42 @@
+# RFC for CI roles extraction
+### An idea to extract common roles for deploying controller from contrail-project-config to be used by Contrail-Windows and DevOps teams.
+
+Author: Katarzyna Rybacka
+
+---
+
+
+## Abstract
+
+Currently, controller5.0 is being deployed in Contrail-Windows CI using roles copied from `contrail-project-config` repository([Link](https://github.com/Juniper/contrail-project-config)).
+There is a need to make this code more generic so it could be used by few teams simultaneously.
+This document specifies the proposed solution to this issue and requests discussion and suggestions for
+improvements.
+
+
+## Introduction
+
+In CI, specifically in `contrail-windows-ci` repository ([Link](https://github.com/Juniper/contrail-windows-ci)) there are three roles `contrail-ansible-deployer`, `kolla-provision-dockers`
+and `yum-repos-prepare` used to deploy controller5.0. The fact that this code is copied from another repository
+could cause some issues e.g. need for backporting code, when any change occures in source code.
+
+
+## Proposal
+
+The idea is to extract generic code used by Contrail-Windows and DevOps teams to another repository
+and make these roles more parameterized. All action steps are proposed to be performed by members of Contrail-Windows team.
+
+
+![img](../pictures/rfc_for_ci_roles_extraction.png)
+
+
+
+### Involved steps
+
+1. Move roles `contrail-ansible-deployer`, `kolla-provision-dockers` and `yum-repos-prepare` to `contrail-zuul-jobs`
+   repository ([Link](https://github.com/Juniper/contrail-zuul-jobs)).
+2. Remove zuul-specific tasks and code from these roles.
+3. Parameterize roles and remove hardcoded parameters by adding variables such as: physical_interface, network_interface.
+4. Add default values to new added variables so this change will be transparent for DevOps team.
+5. Change provisioning of controller5.0 in Contrail-Windows CI to use `contrail-zuul-jobs` repository instead of roles added to `contrail-windows-ci` repository ([Link](https://github.com/Juniper/contrail-windows-ci)) before.
```

Comment:
Just a note - maybe we should consider using `ansible-galaxy` to include roles from `contrail-zuul-jobs`. Then we would maintain current structure of our playbooks, i.e. we `controller` role which includes provisioning roles. Otherwise we would to do something like the following:

- run `vmware-deploy-testenv.yml` playbook
- clone `contrail-zuul-jobs`
- run `provision-controller` playbook

What do you think about this?

Reply:
I have the same opinion that we should consider using ```ansible-galaxy```

Reply:
Also we should consider running these playbooks in containers. Theoretically multiple jobs running in parallel could execute `ansible-galaxy install` with different branch/commit and they will install it to the same directory, i.e. `/var/lib/jenkins/.ansible`.

Reply:
@sodar @PawelMarchewkaCodiLime @katrybacka 

How would you handle making changes to jobs present in galaxy? Versioning, maintaining and CI of it opens a whole new can of worms. 

> Also we should consider running these playbooks in containers

Where would they be stored? Just as Dockerfiles or some repository? This raises similar problems as galaxy. 

How do you solve the problem of coupling playbook to a dockerfile? I can already see all the hacks and env variables present in Dockerfile just so that ansible run within it works... grr..

Reply:
@sodar Actually I believe running playbooks in containers isn't currently necessary. Currently it works without it. Therefore I think it is out of scope here. 

Regarding ```ansible-galaxy```, role can be get directly from github. However I'm not sure how it handles relative path to role inside repository.